### PR TITLE
kara/nonStandardVersionFix #115

### PR DESCRIPTION
**Pull request by kiraarghy**
Added new parseUrl which should be able to parse most unpkg urls we come across as per #115 .

Tested in the following codesandbox https://codesandbox.io/s/jest-test-r3kjb

Should we make it that it only outputs the following object when it has all of those or make some of them nullable? i.e. so you can do `https://unpkg.com/@scope/pkg` not just `https://unpkg.com/@scope/pkg@1.0.0-beta.1/index.js`?

```
{
      name,
      version,
      path,
      file,
    };
```

### CODE REVIEW

File: utils/recursiveDependencyFetch.js [hunk 1]
```diff
@@ -34,9 +35,8 @@ const extractDependencies = (input, pkg) => {
 };
 
 const packageJsonUrl = path => {
-  const [_full, name, version] = path.match(
-    /https:\/\/unpkg.com\/(@?[^@\n]*)@?(\d+\.\d+\.\d+)?/
-  );
+  const { name, version } = parseUrl(path);
+  console.log(`${UNPKG}${name}@${version}/package.json`);
```
Comment (ryan-roemer) [inline utils/recursiveDependencyFetch.js]: Please remove console.log.

File: utils/parseUrl.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    .toLowerCase()
+    .trim()
+    .replace('https://unpkg.com', '')
+    .split('/')
+    .map(part => part.trim())
+    .filter(Boolean);
+  if (parts[0].startsWith('@')) {
+    const nameVersion = parts[1].split('@');
+    return {
+      name: `${parts[0]}/`.concat(nameVersion[0]),
```
Comment (ryan-roemer) [inline utils/parseUrl.js]: NIT: Suggest:

```js
name: `${parts[0]}/${nameVersion[0]}`,
```

for consistency.

File: utils/recursiveDependencyFetch.js [hunk 3]
```diff
@@ -34,9 +35,8 @@ const extractDependencies = (input, pkg) => {
 };
 
 const packageJsonUrl = path => {
-  const [_full, name, version] = path.match(
-    /https:\/\/unpkg.com\/(@?[^@\n]*)@?(\d+\.\d+\.\d+)?/
-  );
+  const { name, version } = parseUrl(path);
```
Comment (ryan-roemer) [inline utils/recursiveDependencyFetch.js]: Good callout for the nullable comment in the description. Right now we return `undefined` implicitly if we don't match something in `parseUrl` and that will break right here.

If we _do_ decide to return undefined we should do something like:

```js
const { name, version } = parseUrl(path) || {};
if (!name) {
  console.error(`Unable to parse path: ${path}`);
  return;
}

**Comment by ryan-roemer** [comment]
Do we have the facility for running pure JS unit tests? A function like `parseUrl` is something that can very easily be tested for all kinds of corner cases and should be :)

File: utils/recursiveDependencyFetch.js (hunk 1 above)
Comment (kiraarghy) [reply utils/recursiveDependencyFetch.js]: oh heck 🤦‍♀

**Comment by kiraarghy** [comment]
> Do we have the facility for running pure JS unit tests? A function like `parseUrl` is something that can very easily be tested for all kinds of corner cases and should be :)

We should be able to do this and exclude files with .test from the now build.

File: package.json [hunk 4]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "prettier": "prettier --config .prettierrc --write \"./**/*.{css,html,json,md,yml}\"",
     "lint": "eslint --config .eslintrc \"./**/*.js\"",
+    "jest": " jest",
```
Comment (VirtualDOMinic) [inline package.json]: is the space in `" jest"` intentional?

File: utils/parseUrl.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      return {
+        name: nameVersion[0],
+        version: nameVersion[1],
+        path: `/${parts.join('/')}`,
+        file: parts.slice(1).join('/'),
+      };
+    }
+  } else {
+    return {
+      name: undefined,
```
Comment (VirtualDOMinic) [inline utils/parseUrl.js]: Following on from the chat on slack, I reckon we should use `null` here

File: utils/recursiveDependencyFetch.js [hunk 6]
```diff
@@ -34,10 +35,13 @@ const extractDependencies = (input, pkg) => {
 };
 
 const packageJsonUrl = path => {
-  const [_full, name, version] = path.match(
-    /https:\/\/unpkg.com\/(@?[^@\n]*)@?(\d+\.\d+\.\d+)?/
-  );
-  return `${UNPKG}${name}@${version}/package.json`;
+  const { name, version } = parseUrl(path) || {};
```
Comment (lukejacksonn) [inline utils/recursiveDependencyFetch.js]: Is this `|| {}` required anymore

File: tests/utils/parseUrl.test.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    expect(parseUrl('https://unpkg.com/')).toEqual({
+      file: null,
+      name: null,
+      path: null,
+      version: null,
+    });
+    expect(parseUrl('https://unpkg.com/@scope/pkg')).toEqual({
+      file: null,
+      name: '@scope/pkg',
+      path: '/@scope/pkg',
```
Comment (lukejacksonn) [inline tests/utils/parseUrl.test.js]: Shouldn't path here be null as technically `@scope/pkg` is just the package name?

File: tests/utils/parseUrl.test.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+const parseUrlFile = require('../../utils/parseUrl');
+
+const parseUrl = parseUrlFile.parseUrl;
+
+describe('parseUrl', () => {
+  it('parses non beta package', () => {
+    expect(parseUrl('https://unpkg.com/pkg-dash@1.0.0/extra.js')).toEqual({
+      name: 'pkg-dash',
+      version: '1.0.0',
+      path: '/pkg-dash@1.0.0/extra.js',
```
Comment (lukejacksonn) [inline tests/utils/parseUrl.test.js]: I always considered the path as the full path minus the package name and the file name. What is everyone else's thoughts on this definition? @alex-saunders @VirtualDOMinic

Comment (kiraarghy) [reply tests/utils/parseUrl.test.js]: I was not sure of this definition, I assumed it meant the unpkg url. 

I assume you mean `https://unpkg.com/@scope/pkg@1.0.0/es6/index.js` file would be: `index.js` and path would be `/es6/index.js`?

Comment (VirtualDOMinic) [reply tests/utils/parseUrl.test.js]: I'd be inclined to agree with @lukejacksonn (and yep @kiraarghy my understanding of it matches yours -- i.e. the current `path` would be the equivalent of `${name}@${version}/${newPath}` where `newPath` is the `path` Luke is suggesting

Comment (lukejacksonn) [reply tests/utils/parseUrl.test.js]: I think we used to use it for telling what folder the resource was in.. so:
```
https://unpkg.com/@scope/pkg@1.0.0/es6/index.js
```
then:
```js
{
  name: '@scope/pkg',
  version: '1.0.0',
  path: 'es6',
  file: 'index.js'
}
```

Then you could reconstruct the URL in full just by combining all the parts. Maybe we could store the full path (like you do here) then create a new key called `folder` or something?

Comment (kiraarghy) [reply tests/utils/parseUrl.test.js]: that's cool with me 😄 requirements are hard.

Comment (lukejacksonn) [reply tests/utils/parseUrl.test.js]: They are, but you are doing great 🙌

Comment (kiraarghy) [reply tests/utils/parseUrl.test.js]: Okay so final structure is such: 

If I called it with: `https://unpkg.com/@scope/pkg@1.0.0/es6/index.js`

```javascript 
{
  name: '@scope/pkg',
  version: '1.0.0',
  directory: 'es6',
  path: /@scope/pkg@1.0.0/es6/index.js
  file: 'index.js'
}
```
